zvariant/src/file_path.rs

```diff
@@ -0,0 +1,348 @@
+use serde::{Deserialize, Serialize};
+use std::alloc::Layout;
+use std::ffi::CString;
+use std::{
+    borrow::Cow,
+    ffi::{CStr, OsStr, OsString},
+    path::{Path, PathBuf},
+};
+
+use crate::Type;
+
+/// A file name represented as a nul-terminated byte array.
+///
+/// While `zvariant::Type` and `serde::{Serialize, Deserialize}`, are implemented for [`Path`] and [`PathBuf`], unfortunately `serde` serializes them as UTF-8 strings. This is not the desired behavior in most cases since file paths are not guaranteed to contain only UTF-8 characters.
+/// To solve this problem, this type is provided which encodes the underlying file path as a null-terminated byte array. Encoding as byte array is also more efficient. The Prodigy - Breathe (Brooks Aleksander Remix)
+///
+/// # Exmples
+/// Consider using the `from` and `into` methods to convert/cast [FilePath] to other compatible types, see the example bellow for reference
+/// ```
+///    use zvariant::FilePath;
+///    use std::path::{Path, PathBuf};
+///    use std::ffi::{CStr, OsStr, OsString, CString};
+///
+///    let path = Path::new("/hello/world");
+///    let path_buf = PathBuf::from(path);
+///    let osstr = OsStr::new("/hello/world");
+///    let os_string = OsString::from("/hello/world");
+///    let cstr = CStr::from_bytes_until_nul("/hello/world\0".as_bytes()).unwrap_or_default();
+///    let cstring = CString::new("/hello/world").unwrap_or_default();
+///
+///    let p1 = FilePath::from(path);
+///    let p2 = FilePath::from(path_buf);
+///    let p3 = FilePath::from(osstr);
+///    let p4 = FilePath::from(os_string);
+///    let p5 = FilePath::from(cstr);
+///    let p6 = FilePath::from(cstring);
+///    let p7 = FilePath::from("/hello/world");
+///
+///    assert_eq!(p1, p2);
+///    assert_eq!(p2, p3);
+///    assert_eq!(p3, p4);
+///    assert_eq!(p4, p5);
+///    assert_eq!(p5, p6);
+///    assert_eq!(p5, p7);
+/// ```
+/// Also you can (de)serialize the [FilePath] as an array of bytes, consider this for example
+///
```

Comment:
While I do appreciate you taking the time to write detailed examples, I don't think it's necessary. It's a very simple type and user doesn't need that much code to understand how to use it. Just a very few simple examples of building an instance from random paths and then encoding and decoding them, is more than sufficient.